### PR TITLE
Check shard limit after applying index templates

### LINKED ISSUE #44567: cluster.max_shards_per_node derives shard count from default not template settings
**Elasticsearch version** (`bin/elasticsearch --version`):
Tested in version 6.x. May hit other versions

**Plugins installed**: []
N/A

**JVM version** (`java -version`):
N/A

**OS version** (`uname -a` if on a Unix-like system):
N/A

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

Have too many shards, and create index or reindex. 

You will receive a warning that this will add [10] shards, irrespective of the number that would be created due to the template for that index setting those shards to a different number.

cluster.max_shards_per_node checks the default number of shards to be created, not the template settings for the index.

**Provide logs (if relevant)**:

### PR DESCRIPTION
**Pull request by jasontedor**
Today when creating an index and checking cluster shard limits, we check the number of shards before applying index templates. At this point, we do not know the actual number of shards that will be used to create the index. In a case when the defaults are used and a template would override, we could be grossly underestimating the number of shards that would be created, and thus incorrectly applying the limits. This commit addresses this by checking the shard limits after applying index templates.

Closes #44567
Relates #34021
Supersedes #44619

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexCreationTaskTests.java [hunk 1]
```diff
@@ -429,7 +429,14 @@ private ClusterState executeTask() throws Exception {
         setupRequest();
         final MetaDataCreateIndexService.IndexCreationTask task = new MetaDataCreateIndexService.IndexCreationTask(
             logger, allocationService, request, listener, indicesService, aliasValidator, xContentRegistry, clusterStateSettings.build(),
-            validator, IndexScopedSettings.DEFAULT_SCOPED_SETTINGS);
+            validator, IndexScopedSettings.DEFAULT_SCOPED_SETTINGS) {
+
+            @Override
+            protected void checkShardLimit(final Settings settings, final ClusterState clusterState) {
```
Comment (gwbrown) [inline server/src/test/java/org/elasticsearch/cluster/metadata/IndexCreationTaskTests.java]: Could we add a short comment here explaining why this is here?

**Review by AthenaEryma — APPROVED** [review]
Thanks for fixing this. LGTM, with one very minor nitpick

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexCreationTaskTests.java (hunk 1 above)
Comment (jasontedor) [reply server/src/test/java/org/elasticsearch/cluster/metadata/IndexCreationTaskTests.java]: I pushed f1ea945a372883f0e201746c0a5aa3312c01645c.

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/2

**Pull request merged**